### PR TITLE
CPUBlas: Use opmath_type for alpha/beta scalars

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80977
* #65840
* __->__ #65839

RUN_TORCHBENCH: ALL

### CODE REVIEW

File: aten/src/ATen/native/cpu/BlasKernel.cpp [hunk 1]
```diff
@@ -73,7 +73,7 @@ void gemm_transa_(
     const scalar_t *b_ = b;
     for (int64_t j = 0; j < n; j++)
     {
-      scalar_t sum = 0;
+      opmath_t sum = 0;
       for(int64_t l = 0; l < k; l++)
         sum += a_[l]*b_[l];
```
Comment (Lezcano) [inline aten/src/ATen/native/cpu/BlasKernel.cpp:78]: All the other operaitons in this PR happen in `opmath_t` but this one. Should at least one of the arguments be casted to `opmath_t` for the whole operation to happen in `opmath_t`?

**Review by lezcano — COMMENTED** [review]
Just left a small comment. Also, just to see if I got this right. The operations are performed in single precision but read in half precision because this is not computation bound but memory bound. Is that it?

**Comment by peterbell10** [comment]
>The operations are performed in single precision but read in half precision because this is not computation bound but memory bound. Is that it?

`Half` and `BFloat16` on CPU work by converting to `float`, doing the operation, and then converting back to the original type. So keeping intermediates in `float` precision should actually perform better, as well as being more precise.

**Comment by ngimel** [comment]
Maybe we should disable fp16 matmul on cpu because it produces inaccurate results and is very slow #69969. Even with this PR, `transa` function will now be accurate, but `transb`, since it accumulates directly in output, will still truncate after each addition.

**Comment by peterbell10** [comment]
@ngimel I see that half support was removed in #75647, but this should still be relevant for BFloat16. AFAIK there is no reason not to make this change, even if truncation will still take place elsewhere in the calculation. Even without any precision improvements, performing the calculation in `opmath_t` should perform better since we don't convert intermediate values back to BFloat16 repeatedly.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**